### ISSUE TITLE
[arm] Assertion failed '!"Called ConstantValue<T>(vn), but type(T) != type(vn); Use CoercedConstantValue instead."'

### ISSUE THREAD

**Issue by VincentBu** [dotnet/coreclr#23538]
**Job:**
https://ci.dot.net/job/dotnet_coreclr/job/master/view/arm/job/jitstress/job/arm_cross_checked_windows_nt_corefx_baseline_tst/60/

**Log:**
```
C:\j\workspace\arm_cross_che---eb7b805d\_\fx\artifacts\bin\tests\System.Utf8String.Experimental.Tests\netcoreapp-Windows_NT-Release-arm>set COMPlus_TieredCompilation=0 

C:\j\workspace\arm_cross_che---eb7b805d\_\fx\artifacts\bin\tests\System.Utf8String.Experimental.Tests\netcoreapp-Windows_NT-Release-arm>C:\j\workspace\arm_cross_che---eb7b805d\_\fx\artifacts\bin\testhost\netcoreapp-Windows_NT-Release-arm\dotnet.exe xunit.console.dll System.Utf8String.Experimental.Tests.dll -xml testResults.xml -nologo -nocolor -notrait category=nonnetcoreapptests -notrait category=nonwindowstests -notrait category=IgnoreForCI -notrait category=failing -notrait category=OuterLoop 
  Discovering: System.Utf8String.Experimental.Tests (method display = ClassAndMethod, method display options = None)
  Discovered:  System.Utf8String.Experimental.Tests (found 92 test cases)
  Starting:    System.Utf8String.Experimental.Tests (parallel test collections = on, max threads = 8)

Assert failure(PID 10740 [0x000029f4], Thread: 10484 [0x28f4]): Assertion failed '!"Called ConstantValue<T>(vn), but type(T) != type(vn); Use CoercedConstantValue instead."' in 'System.Tests.Utf8ExtensionsTests:AsBytes_FromSpan_Default():this' (IL size 122)

    File: d:\j\workspace\arm_cross_che---99792267\src\jit\valuenum.h Line: 810
    Image: C:\j\workspace\arm_cross_che---eb7b805d\_\fx\artifacts\bin\testhost\netcoreapp-Windows_NT-Release-arm\dotnet.exe

 exit code -1073740286
```

**Comment by BruceForstall** [comment]
@dotnet/jit-contrib

**Comment by BruceForstall** [comment]
Call stack:
```
 # Child-SP RetAddr  : Args to Child                       : Call Site
00 (Inline) -------- : -------- -------- -------- -------- : KERNELBASE!DebugBreak [minkernel\kernelbase\debug.c @ 143] 
01 05377548 57c92884 : 053795a0 57c92885 57f316d0 0a02ba0c : KERNELBASE!wil::details::DebugBreak+0x6 [internal\sdk\inc\wil\resultmacros.h @ 1837] 
*** WARNING: Unable to verify checksum for C:\brucefo\fxruntime\shared\Microsoft.NETCore.App\9.9.9\clrjit.dll
02 05377550 57c2e4fa : 57f316d0 0a02ba0c 0000007a 57cc456f : clrjit!assertAbort+0xf8 [f:\gh\coreclr10\src\jit\error.cpp @ 310] 
03 053795b8 57c2e0c0 : 00000001 0ab04588 00000001 00000007 : clrjit!ValueNumStore::ConstantValueInternal<unsigned int>+0x156 [f:\gh\coreclr10\src\jit\valuenum.h @ 814] 
04 05379600 57df3ef0 : 0ab04500 e0452259 05379738 57df3ef1 : clrjit!ValueNumStore::ConstantValue<unsigned int>+0x18 [f:\gh\coreclr10\src\jit\valuenum.h @ 828] 
05 05379618 57e0582a : 0a02aeac 00000004 01070e0e 00000040 : clrjit!ValueNumStore::EvalFuncForConstantArgs+0x3a0 [f:\gh\coreclr10\src\jit\valuenum.cpp @ 2780] 
06 05379750 57e07cdc : 00000081 57c32000 000e0e01 01010100 : clrjit!ValueNumStore::VNForFunc+0x252 [f:\gh\coreclr10\src\jit\valuenum.cpp @ 2079] 
07 053797f0 57e10a72 : 00000081 00000041 00000081 00000081 : clrjit!ValueNumStore::VNPairForFunc+0x30 [f:\gh\coreclr10\src\jit\valuenum.h @ 499] 
08 05379830 57e0b0aa : 00000081 00000081 00000081 00000081 : clrjit!Compiler::fgValueNumberTree+0x2f76 [f:\gh\coreclr10\src\jit\valuenum.cpp @ 8188] 
09 0537a188 57e08c6a : 00000309 0000001c 00000001 00000005 : clrjit!Compiler::fgValueNumberBlock+0x892 [f:\gh\coreclr10\src\jit\valuenum.cpp @ 6150] 
0a 0537a338 57c73a8c : 0537010e 0a8de000 000000b2 0137a301 : clrjit!Compiler::fgValueNumber+0x45a [f:\gh\coreclr10\src\jit\valuenum.cpp @ 5891] 
0b 0537a408 57c76162 : 01010101 ff010101 05370000 00000000 : clrjit!Compiler::compCompile+0x694 [f:\gh\coreclr10\src\jit\compiler.cpp @ 4687] 
```

**Comment by briansull** [comment]
Fixed with #23728
